### ISSUE TITLE
Scheduler won't start if --zk flag contains at least one unreachable Zookeeper node

### ISSUE THREAD

**Issue by serejja** [mesos/kafka#141]
Using latest mesos/kafka (489f57cab1a8328d944a7a4a4ab8782cd9d9abd6 to be precise) I'm getting into the situation when scheduler cannot start if `--zk` flag contains at least one unreachable node.

Minimal example:
1. `vagrant up` using provided VMs.
2. build the framework and place kafka tgz to resolve dependencies properly. 
3. `./kafka-mesos.sh scheduler --debug true --master=master:5050 --zk=master:2181,non-existing-dns-name:2181 --storage=zk:/mesos-kafka`

This only happens when --zk contains a DNS name that cannot be resolved (e.g. if I add `slave0:2181` to the list which resolves fine but does not have Zookeeper there - it does not crash but warns that connection is refused) AND when you use `--storage=zk:/whatever`.

This way if you use either mesos-dns or mesos-consul for service discovery once one of your Zookeeper nodes dies and DNS entry does not exist anymore the scheduler crashes and is not able to restart (if run via Marathon)

Exception:

```
Exception in thread "main" org.I0Itec.zkclient.exception.ZkException: Unable to connect to master:2181,non-existing-dns-name:2181
    at org.I0Itec.zkclient.ZkConnection.connect(ZkConnection.java:66)
    at org.I0Itec.zkclient.ZkClient.connect(ZkClient.java:876)
    at org.I0Itec.zkclient.ZkClient.<init>(ZkClient.java:98)
    at org.I0Itec.zkclient.ZkClient.<init>(ZkClient.java:84)
    at ly.stealth.mesos.kafka.Cluster$ZkStorage.zkClient(Cluster.scala:132)
    at ly.stealth.mesos.kafka.Cluster$ZkStorage.loadJson(Cluster.scala:147)
    at ly.stealth.mesos.kafka.Cluster$Storage.load(Cluster.scala:95)
    at ly.stealth.mesos.kafka.Cluster.load(Cluster.scala:53)
    at ly.stealth.mesos.kafka.Scheduler$.start(Scheduler.scala:331)
    at ly.stealth.mesos.kafka.Cli$SchedulerCli$.handle(Cli.scala:369)
    at ly.stealth.mesos.kafka.Cli$.exec(Cli.scala:54)
    at ly.stealth.mesos.kafka.Cli$.main(Cli.scala:36)
    at ly.stealth.mesos.kafka.Cli.main(Cli.scala)
Caused by: java.net.UnknownHostException: non-existing-dns-name
    at java.net.Inet4AddressImpl.lookupAllHostAddr(Native Method)
    at java.net.InetAddress$1.lookupAllHostAddr(InetAddress.java:922)
    at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1314)
    at java.net.InetAddress.getAllByName0(InetAddress.java:1267)
    at java.net.InetAddress.getAllByName(InetAddress.java:1183)
    at java.net.InetAddress.getAllByName(InetAddress.java:1119)
    at org.apache.zookeeper.ClientCnxn.<init>(ClientCnxn.java:387)
    at org.apache.zookeeper.ClientCnxn.<init>(ClientCnxn.java:332)
    at org.apache.zookeeper.ZooKeeper.<init>(ZooKeeper.java:383)
    at org.I0Itec.zkclient.ZkConnection.connect(ZkConnection.java:64)
    ... 12 more
```

**Comment by dmitrypekar** [comment]
Digging deeper into this shown that Kafka 0.8.2.1 has the same problem. If zookeeper.connect is configured to have unresolveable hostname, kafka startup fails like this:

```
[2015-11-06 14:25:39,743] FATAL Fatal error during KafkaServerStartable startup. Prepare to shutdown (kafka.server.KafkaServerStartable)
org.I0Itec.zkclient.exception.ZkException: Unable to connect to master:2181,unknown:2181
    at org.I0Itec.zkclient.ZkConnection.connect(ZkConnection.java:66)
    at org.I0Itec.zkclient.ZkClient.connect(ZkClient.java:876)
    at org.I0Itec.zkclient.ZkClient.<init>(ZkClient.java:98)
    at org.I0Itec.zkclient.ZkClient.<init>(ZkClient.java:84)
    at kafka.server.KafkaServer.initZk(KafkaServer.scala:157)
    at kafka.server.KafkaServer.startup(KafkaServer.scala:82)
    at kafka.server.KafkaServerStartable.startup(KafkaServerStartable.scala:29)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at ly.stealth.mesos.kafka.KafkaServer.start(BrokerServer.scala:49)
    at ly.stealth.mesos.kafka.Executor$.ly$stealth$mesos$kafka$Executor$$runBroker0$1(Executor.scala:76)
    at ly.stealth.mesos.kafka.Executor$$anon$1.run(Executor.scala:98)
Caused by: java.net.UnknownHostException: unknown
    at java.net.Inet4AddressImpl.lookupAllHostAddr(Native Method)
    at java.net.InetAddress$1.lookupAllHostAddr(InetAddress.java:922)
    at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1316)
    at java.net.InetAddress.getAllByName0(InetAddress.java:1269)
    at java.net.InetAddress.getAllByName(InetAddress.java:1185)
    at java.net.InetAddress.getAllByName(InetAddress.java:1119)
    at org.apache.zookeeper.client.StaticHostProvider.<init>(StaticHostProvider.java:61)
    at org.apache.zookeeper.ZooKeeper.<init>(ZooKeeper.java:445)
    at org.apache.zookeeper.ZooKeeper.<init>(ZooKeeper.java:380)
    at org.I0Itec.zkclient.ZkConnection.connect(ZkConnection.java:64)
    ... 13 more
```

Taking this into account I think it would be still not possible to use such zk.connect, even if fixed somehow for the Scheduler itself (we would just not be able to start Kafka broker).

This seems to be a bug of `zkclient-0.3.jar` package. Tested on `zkclient-0.6.jar` - same issue.

**Comment by dmitrypekar** [comment]
Waiting 10 days for the decision from the ticket owner. If no additional info, will close the ticket as 'wan't fix'

**Comment by serejja** [comment]
Closing as it doesn't seem to be an issue in mesos/kafka but rather in Zookeeper client itself. 

Thanks for investigating

**Comment by poths** [comment]
Note: This issue seems to have been fixed in Zookeeper 3.5.0 (see https://issues.apache.org/jira/browse/ZOOKEEPER-1576), but I do not know when Zookeeper 3.5 will be released and Kafka can upgrade to the new Zookeeper.
